Add option to disable Extended Master Secret

tmshort opened the pull request:
Add SSL_OP_NO_EXTENDED_MASTER_SECRET, that can be set on either
an SSL or an SSL_CTX. When processing a ClientHello, if this flag
is set, do not indicate that the EMS TLS extension was received in
either the ssl3 object or the SSL_SESSION. 

Add a commit to increase SSL options to be 64 bits on all platforms.
<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING

Other than that, provide a description above this comment if there isn't one already

If this fixes a github issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] documentation is added or updated
- [x] tests are added or updated

kroeckx commented:
Can you explain the use case? I could maybe understand that a client might not want to send that extension because it might break some other implementation, but I see no point in the server not using it when the client supports it.

tmshort commented:
Yes, it's a broken server.
I see your point, on the other hand, I don't see a reason not to offer feature parity server/client where possible (it's also a fairly small change, which doesn't help either argument).

davidben commented:
Out of curiosity, what is the server and how is it broken? (EMS-specific, or is this extension intolerance?) We ship EMS in all our clients with no way to disable it. I don't recall seeing evidence of compatibility problems.

tmshort commented:
Specifically, it's **IBM HTTP Server for WebSphere Application Server, Version  8.5.5.7**, which runs on top of Java. If we remove the EMS extension, it works; but reordering extensions might also work (but that's harder to do in 1.1.0 vs master due to code changes), and it could be related to a server not handling empty extensions at the end of the extension list correctly.

A possibly similar problem was reported on NSS:

https://bugzilla.mozilla.org/show_bug.cgi?id=1243641  

Which reordered the extensions. Disabling EMS was the easiest solution vs. reordering extensions.

There's also #2669, which may be related.

kroeckx commented:
I think you mean #2691. The other seems to be about encrypt then mac?

tmshort commented:
Definitely related to #2691, possibly related to #2669.

kroeckx commented:
I guess I prefer a solution that tries to avoid the problem, instead of having to disable it.

tmshort commented:
@kroeckx, I agree. I'm looking into that now, but I have to wait for the test server to be setup.

That being said, we have use for disabling EMS; and there's also an option to disable EtM, which has similar arguments for/against the ability to disable.

tmshort commented:
And I just found this on the IETF mailing list:

https://www.ietf.org/mail-archive/web/tls/current/msg19719.html

So, I will be looking at moving EMS, EtM and SessionTickets to not be last. As pointed above, this is easy on master, harder on 1.1.0.

tmshort commented:
Further investigation has shown that WebSphere 8.5.5 cannot handle EtM, so porting the SSL_OP_NO_ENCRYPT_THEN_MAC option from master to 1.1.0 might be a good idea.

kaduk commented:
> backporting SSL_OP_NO_ENCRYPT_THEN_MAC

#3922 FWIW

mattcaswell commented:
Is this resolved now by moving the sig algs extension?

tmshort commented:
@mattcaswell the fundamental issue of WebSphere is fixed with the sigalgs changes. However, being able to disable EMS might still be a useful feature.

kaduk commented:
@mattcaswell do you think we should leave this open or close it after the sigalg extension was moved?

tmshort commented:
I think the change still has merit, regardless of the sigalg commit. But I'm not going to force the issue.

mattcaswell commented on ssl/statem/extensions_clnt.c:
```diff
@@ -1408,6 +1411,8 @@ int tls_parse_stoc_etm(SSL *s, PACKET *pkt, unsigned int context, X509 *x,
                        size_t chainidx, int *al)
 {
+    if (s->options & SSL_OP_NO_EXTENDED_MASTER_SECRET)
+        return 1;
```
I'm not sure if this change is necessary. If the option is set then we will not have proposed EMS. If the server sends an unsolicited ems response then the extensions system should barf. So we should only ever get here if we proposed it - which means the option is not set.

tmshort replied:
Correct. This is just safety, but not strictly necessary.

tmshort replied:
I will point out that tls_parse_stoc_etm() - immediately above this function - does check for SSL_OP_NO_ENCRYPT_THEN_MAC. This behavior should fall under the same umbrella.

mattcaswell commented on test/sslapitest.c:
```diff
@@ -481,8 +481,53 @@ static int test_early_cb(void)
     return testresult;
 }
+
+static int test_no_ems(void)
```
Doesn't this function need to set the max protocol to TLSv1.2?

tmshort replied:
Yes.

tmshort commented:
Rebased

kaduk commented:
There are some setups such as those described in https://tools.ietf.org/html/draft-erb-lurk-rsalg-01 that are not easily adopted to be compatible with EMS.

kaduk commented:
ping @mattcaswell ?

indutny commented:
Is there anything else that needs to be improved in this PR before landing it?

kaduk commented:
Hopefully Matt can take another look at some point.

mattcaswell commented:
Due to the passage of time, we now have a new problem with this PR. Namely that we have run out of free `SSL_OP_*` numbers.

If the next release is 1.2.0 then there are a number of values marked as "reserved until 1.2.0" which could be used. It seems likely based on discussions so far that the next release will indeed by 1.2.0, but that decision has not yet been formally agreed by the OMC.

tmshort commented:
Alternatively, SSL_CTX_set_options_ex(), et al, could be developed, but I'm not sure that's worth the effort.
If the next significant version does become 1.2.0, would it make sense to extend (or at least fix) the options size to 64bits?

mattcaswell commented:
Probably, yes.

tmshort commented:
Added code to extend options to 64-bits. This is only needed if we choose not to reuse older bits and/or want to provide more than 32 option bits (since we kinda-sorta ran out of bits in 1.1.1).
Because of the lack of bits, this cannot be committed until 1.2.0

richsalz commented:
A more portable method might be to use character strings or something like FDSET

tmshort commented:
I just extended the existing API. Indeed, there are many possible solutions to extending the options interface, but those might break source compatibility. The SSL_CONF interface does use strings, but ends up going through this option interface.

tmshort commented:
(I suppose that OpenSSL 2.0 would be the version that says to-hell with source compatibility...)

davidben commented:
You all could also just add a C function...

tmshort commented:
That is also true, but that means losing out on the (easy) SSL_CONF integration.

richsalz commented:
No, the conf code would have to call the function.  A bit more work for the openssl developer, no extra work for the config user.

tmshort commented:
@richsalz that's what I meant!

richsalz commented:
It's trivial:
```
if (strcasecmp(word, "noetm")) call_new_function()
else set_flag
```
inside the for loop.

tmshort commented:
I was trying to avoid adding extra APIs and preferred to use existing mechanisms.
From the pushback, it almost sounds as though people don't want to extend the option field size?

mattcaswell commented:
I think extending the option field size makes perfect sense.

richsalz commented:
I thought the issue was that we are already using 32bits. Not sure how you extend the field without one of new API's, using strings, or using FDSET's.

tmshort commented:
In 1.2, the API can be updated to use 64-bit, but not before then. If the next release is 1.2, then it's something that could be done. If the next release is not 1.2, then some other mechanism would have to be used. So, it depends on what the next release is.

richsalz commented:
More importantly, it also depends on if the OMC is willing to give up 32bit platforms in the next non-compatible release.

mattcaswell commented:
It does not require us to give up 32bit platforms. It only requires that a platform supports a 64 bit unsigned int type. Which we already have as a requirement in 1.1.0+.

richsalz commented:
Thanks for the correction.

tmshort commented:
@richsalz It would require an source-incompatible change; which means that it cannot be done within 1.1.x timeline, but would require a "significant" release (avoiding the terms "major" and minor" due to the ongoing conversation on semantic versioning)

And as per other conversations, I want to avoid exploding APIs.

tmshort commented:
Then again, if the reserved bits become available, the API change becomes moot.

kumarrishav commented:
Will this feature land in Node 12?

tmshort commented:
I'm not sure what you're asking? @kumarrishav

kumarrishav commented:
@tmshort So, i was talking in context of https://github.com/openssl/openssl/issues/7246. Session resume was breaking in node 10. So, i was thinking, this feature would fix the issue .

davidben commented:
Can you please describe how exactly EMS is breaking Node? #7246 is asking for something invalid.

kumarrishav commented:
Hmm, It's not exactly breaking node but the way session resumption is build in infrastructure. So,  #7246 shows, that how first session-resumption fails.  Fedor is the right guy who can explain more about this.

davidben commented:
But that is to be expected. The same behavior would happen if you upgraded to add TLS 1.3 support. Your old TLS 1.2 session would be rejected since, naturally, one cannot resume a TLS 1.2 session at TLS 1.3.

tmshort commented:
ping @mattcaswell ?

dwmw2 commented:
As noted in https://github.com/openssl/openssl/issues/7246 my use case for session resumption is Cisco AnyConnect VPN, where we "resume" a session that never really existed in the first place. That synthesised session didn't have "extended master secret".

When OpenSSL sets the extended_master_secret extension in its ClientHello, the server MUST NOT allow the session resume to complete. And that means we just get a connection failure instead.

I could live with `SSL_OP_NO_EXTENDED_MASTER_SECRET` if it's backported to every version of OpenSSL which sends the unwanted extension even for resuming a session that can't use it. But I'd much rather OpenSSL just got it right for itself, based on the session being resumed.

tmshort commented:
Unfortunately, previous versions of OpenSSL (1.1.1) have used up all the SSL_OP bits, so backporting may not be feasible.

dwmw2 commented:
Leaving us with just the "DTRT" option? Shame... :)

tmshort commented:
There's whole discussion above about SSL_OP vs. new API. 1.1.0 is only getting security fixes now, and 1.0.2 is on its last legs as well, as both are out of support by the end of this year. So, I don't know if it would be acceptable to backport a new API. @mattcaswell ?

tmshort commented:
> But I'd much rather OpenSSL just got it right for itself, based on the session being resumed.

This shouldn't be too hard of a change, but I don't have time right now to do it...

dwmw2 commented:
For now I think I just need to disable the DTLSv1.2 support in the OpenConnect client when it's built with OpenSSL >= 1.1.0? There's no way for it to work?

tmshort commented:
I don't know what Cisco has done w.r.t. DLTSv1.2 AC support; I left 4 years ago. I didn't think it was on their radar, since they still used the pre-RFC version.

dwmw2 commented:
As of v9.10, The ASAv virtual appliance can apparently now support DTLSv1.2. Having the data plane implemented in hardware means that real ASAs will be slower to follow, of course. The latest clients do support it. They use OpenSSL 1.0.2 so they don't trip over this issue... yet :)

dwmw2 commented:
Wait... there *are* flags. They are already in the session ASN.1. They are already *not* set in the session I've synthesised to be resumed. It's just a plain bug. Won't this fix it?

```
--- a/ssl/t1_lib.c
+++ b/ssl/t1_lib.c
@@ -1395,15 +1395,20 @@ unsigned char *ssl_add_clienthello_tlsext(SSL *s, unsigned char *buf,
     }
 #endif
 
-    /*-
-     * check for enough space.
-     * 4 bytes for the EMS type and extension length
+    /*
+     * Don't advertise extended master secret if we are resuming a
+     * session which didn't have it.
      */
-    if (CHECKLEN(ret, 4, limit))
-        return NULL;
-    s2n(TLSEXT_TYPE_extended_master_secret, ret);
-    s2n(0, ret);
-
+    if (s->new_session || !s->session || (s->session->flags & SSL_SESS_FLAG_EXTMS)) {
+        /*
+         * check for enough space.
+         * 4 bytes for the EMS type and extension length
+         */
+        if (CHECKLEN(ret, 4, limit))
+            return NULL;
+        s2n(TLSEXT_TYPE_extended_master_secret, ret);
+        s2n(0, ret);
+    }
     /*
      * WebSphere application server can not handle having the
      * last extension be 0-length (e.g. EMS, EtM), so keep those
```

davidben commented:
> But I'd much rather OpenSSL just got it right for itself, based on the session being resumed.

No, that patch is insecure and must not be accepted. This means that session resumption causes a "stickiness" that an EMS-supporting client and server from discovering that EMS should be enabled. It's the same kind of version-stickiness problem that OpenSSL's older version negotiation implementation had. (That problem meant that some people couldn't disable SSL 3.0 despite all applicable clients supporting TLS!)

tmshort commented:
In @dwmw2 use-case, the session must always be resumed. It cannot be forced into a new handshake (the server will fail instead). So while it might be nice to keep the resumed session parameters the same from connection to connection, the majority use-case is to allow the server to do a full handshake if it wants to use new parameters.
This leads to the need for EMS to be configurable per-connection (like EtM).

davidben commented:
Right, to be clear, I'm saying not to merge @dwmw2's patch that auto-detects, not this PR. I don't have strong opinions about adding an off-by-default option as in this PR.

Although, per #7246, note that undeclinable sessions are thoroughly outside the usual semantics of TLS session resumption. Protocols should not be making such assumptions. Historically, OpenSSL has made changes that were only backwards compatible given declinable sessions.

dwmw2 commented:
For reference, the auto-detect patch is now https://github.com/openssl/openssl/pull/7999

In the "must resume" use case, I am also constructing the session from scratch. So perhaps instead of simply *refraining* from setting the `SSL_SESS_FLAG_EXTMS` flag, I could set a new flag which prevents extms for that session instead?

dwmw2 commented:
Of course, in the Cisco case it's kind of bizarre to be fretting over extms vs. standard derivation given that the master secret isn't being derived that way anyway; the client pulled it out of /dev/random.

We might as well pretend that it *is* an extms session that's being "resumed". Perhaps the server should look at whether the TCP connection supports extms, and then generate a DTLS session which matches. Or the server should have an option to ignore the extms-matching status and accept the connection anyway. Or the server application could add *two* sessions to its "cache"; one with extms and one without, with the same session-id and master secret. Or something...

Suggestions welcome. I'm just trying to achieve maximal compatibility with what Cisco have already decided to put into the field. Ideally while allowing users/distributions to build the client against any currently supported version of OpenSSL.

_(Of course, the ideal solution is not to do session resumption at all, and to use PSK as $DEITY intended. We do that in the open source client and server implementations, as an extension to Cisco's original protocol. But when they updated to DTLSv1.2 they elected not to do it that way)_

tmshort commented:
Again, I left Cisco 4 years ago... I have no idea what they're doing!

tmshort commented:
I rebased this on top of master, and modified the option functions to be 64-bit extensions of existing functions. This can be backported to 1.1.1 cleanly (with the exception of libssl.num).

tmshort commented:
Fix casting.
I admit this is not an optimal solution for master/3.0.0, where it would be better to just covert the options functions to be 64-bit.

tmshort commented:
ping @openssl 
This is setup to be backported to 1.1.1. But if we don;t care about backporting, then we can just change the option APIs to use uint64_t, and get rid of the 64-bit suffixed versions.

mattcaswell commented:
I'm thinking that there is no pressing reason to backport this - so we could remove the 64 bit suffixed versions. Incidentally, if we only target master then this actually frees up some SSL_OP_ values because there are some which are unused in 1.1.1 but reserved for ABI compatibility reasons until the next major release (which 3.0.0 is). We should still convert the functions to take uint64_t though anyway.

tmshort commented:
Done @mattcaswell

mattcaswell commented on include/openssl/ssl.h:
```diff
@@ -297,23 +297,27 @@ typedef int (*SSL_verify_cb)(int preverify_ok, X509_STORE_CTX *x509_ctx);
+# define SSL_OP_NO_EXTENDED_MASTER_SECRET                0x00000001U
+
+/* Reserved value (until OpenSSL 3.0.0)                  0x00000001U */
+/* Reserved value (until OpenSSL 3.0.0)                  0x00000002U */
```
Why not just delete all these reserved values - since they are no longer reserved, i.e. master will become OpenSSL 3.0.0.

kaduk replied:
It seems helpful to have annotations to remind anyone preparing a change for backport of the constraints involved

mattcaswell commented on include/openssl/ssl.h:
```diff
@@ -586,12 +590,12 @@ typedef int (*SSL_async_callback_fn)(SSL *s, void *arg);
+uint64_t SSL_CTX_clear_options(SSL_CTX *ctx, uint64_t op);
+uint64_t SSL_clear_options(SSL *s, uint64_t op);
+uint64_t SSL_CTX_set_options(SSL_CTX *ctx, uint64_t op);
+uint64_t SSL_set_options(SSL *s, uint64_t op);
```
I wonder if these signature changes will result in any compilation issues for end user applications.

tmshort replied:
For (most) 64-bit platforms, it shouldn't make much of a difference; assuming **unsigned long** ~ **uint64_t**
For 32-bit platforms, there will be some issues, and users would have to cast them away, or handle the return value properly. Since all the values are in the lower 32-bits, casting won't have issues until more options are used.

kaduk replied:
It's unclear that this counts as "API compatible; e.g., I expect modern compilers to warn about the truncated return value.

tmshort replied:
Yes, compiles will warn, but it won't break anything until upper-32-bit option values are defined.

mattcaswell replied:
I think this warrants a discussion on openssl-project

mattcaswell replied:
I started to write an email to the list about this, but in putting it together I think I answered the question I was going to ask. Specificially openssl/web#82 says:

"No existing public interface can be modified except where changes are unlikely to break source compatibility."

On thinking about this I think a warning is likely to be sufficient to break some builds. Many builds treat warnings as errors. So I'm thinking that this change is probably a step too far.

My suggestion is to remove it from this PR - since really this is just an incidental tangent to the main point of this PR. If we are only targetting 3.0 then we can reuse one of the currently "reserved" SSL_OP_ values for now. We should probably also raise an issue that we're running out of SSL_OP_ values and we can figure out in a separate PR the best way of handling it.

tmshort replied:
Thanks (I should've come back to look at this earlier). Certainly have a 64-suffix version would allow for extension. This would only be an issue with 32-bit platforms. Let's say we drop those. ;) /s

tmshort replied:
Indeed, if we are only targeting 3.0.0, then we are fine re-using the bit. But, I do see a use for adding it to 1.1.1, but not critically.

mattcaswell replied:
> This would only be an issue with 32-bit platforms. Let's say we drop those.

Technically those with a 32-bit long. This includes 64-bit Windows. :-(

tmshort replied:
I was being sarcastic! :) But I undid the 64-bit changes.

mattcaswell commented on test/sslapitest.c:
```diff
@@ -571,6 +571,53 @@ static int test_client_hello_cb(void)
+
+    SSL_CTX_set_options(sctx, SSL_OP_NO_EXTENDED_MASTER_SECRET);
+    SSL_CTX_set_min_proto_version(sctx, TLS1_VERSION);
+    SSL_CTX_set_max_proto_version(sctx, TLS1_2_VERSION);
```
This could be done more easily in the arguments to `create_ssl_ctx_pair`

tmshort commented:
The build failure appears to be related to the sparc_array implementation.

mattcaswell commented:
````
$ ./config -d --strict-warnings no-tls1_2
$ make
$ make test
...
Test Summary Report
-------------------
../test/recipes/80-test_ssl_new.t                (Wstat: 256 Tests: 30 Failed: 1)
  Failed test:  30
  Non-zero exit status: 1
Files=155, Tests=1369, 130 wallclock secs ( 1.00 usr  0.22 sys + 131.60 cusr  8.01 csys = 140.83 CPU)
Result: FAIL

tmshort commented:
The last commit should fix the test issues.

mattcaswell commented:
CI red crosses are unrelated.

mattcaswell commented:
Needs 2nd review.

@openssl/committers

mattcaswell commented:
Pushed. Thanks.

dwmw2 commented:
Did we conclude that we can't have `SSP_OP_NO_EXTENDED_MASTER_SECRET` in OpenSSL 1.1.x? 

I have used it in OpenConnect [(here)](http://git.infradead.org/users/dwmw2/openconnect.git/commitdiff/323849bab81390b5076ab39afae8d902ebb3348f) to avoid asking for extms when resuming the (fake) DTLS session. But that's only for OpenSSL HEAD; with OpenSSL 1.1.0 and 1.1.1 it's still requesting it in the ClientHello. The only reason I get away with this for now is because neither Cisco's server nor the free ocserv implementation actually support it. If they did, they should refuse to resume the session as discussed.

mattcaswell commented:
> Did we conclude that we can't have SSP_OP_NO_EXTENDED_MASTER_SECRET in OpenSSL 1.1.x?

Yes, that was the decision.

dwmw2 commented:
Thanks. Are we open to... less elegant solutions such as perhaps never allowing extms for `DTLS1_BAD_VER`? 

That's only a partial solution though; Cisco have now finally updated to support a standard version of DTLS (v1.2) at least in the virtual ASA applicances. And they're still using the session resume hack, instead of proper negotiation with PSK, as we do in the open ocserv server.

Not quite sure how to address that, other than relying on the *server* not to accept extms even if we do offer it. We are manually generating the ASN.1 session state so perhaps we could put something special in there like a `SSL_SESS_FLAG_DISABLE_EXTMS` flag, coupled with a patch along the lines of https://github.com/openssl/openssl/pull/7999/commits/02d405d9744d87a9be8f14fd13f894d16b8f3394

That wouldn't have the problems of the original patch because `SSL_SESS_FLAG_DISABLE_EXTMS` would be an explicit instruction from the client side, equivalent to `SSL_OP_NO_EXTENDED_MASTER_SECRET`. We *wouldn't* refuse to do extms for the general case of resuming a non-extms session, in the case where the client/server have been upgraded and can now support it.

Not sure I like it though. Better ideas?

chipitsine commented:
Hello, I have some concern about it.

we have a lot of customers with CryptoPro 4.0R3 (it is russian gost implementation). CryptoPro 4.0R3 does not like EMS in server helo handshake.

so, what are we supposed to do with `SSP_OP_NO_EXTENDED_MASTER_SECRET` not being ported to openssl-1.1.1 ?

are we supposed to 

1. wait for 3.0 release ?
2. install openssl master ?


I mean that not being ported to 1.1.1 branch is weird and left us no choice for today.
I'm open to your suggestions

beldmit commented:
From my experience, I can confirm that some GOST implementations do not support EMS correctly, so the option (disabled by default, of cause) would be rather useful.
